keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+1. If the entry point to the application handles some Kubernetes object like pod or node this object should also be associated with the entry point Context.
+1. Starting from the identified entry points Context containing a logger should be propagated to all places where we log anything. The propagation should be done directly by passing Context as an argument to called functions/methods or indirectly by associating a Context with existing objects passed between functions/methods. In all places where context has been added references to module logger should be replaced with `k8s.io/component-base/kontext.getLogger(context)` calls.
+
+Manual logs verification
+
+1. Detailed documentation for the structured logging best practices in Kubernetes and migration guide for moving from unstructured to structured logs should be prepared.
+1. For each automatically generates logr call following checks should be performed manually:
+    * If log does not provide true value from the developer or user point of view it should be removed.
+    * Log message should be checked if it is meaningful.
+    * Log level should be verified if it correct (Error() vs Info()).
+    * Log verbosity should be verified if it correct..
+    * Log entry fields should be added or changed.
+    * Where it makes sense Kubernetes objects should be attached to context.
+
+#### Stage 3 / GA
+
+1. All places were we log anything in the Kubernetes should be reviewed before reaching this stage and all log messages and fields should be stage enough so that they should not change significantly in the next releases.
+1. `json` log format should become a default.
+
+#### Stage 4 / Deprecation
+
+Cleanup
+
+1. Support for `legacy` logging format should be removed.
+1. klog wrapper package and all of its references should be removed from Kubernetes source code.
+1. All dependencies which are included in Kubernetes and which are currently using klog for logging should be migrated to logr. After this is done klog to logr bridge using klogr should be removed from Kubernetes source code.
+
+
+### User Stories
+
+#### Story 1 - Kubernetes developer working with local logs
+
+TODO(serathius):No major impact when switching from legacy to json format
+
+#### Story 2 - Cluster Administrator configures log ingestion into ElasticSearch
+
+TODO(serathius): Show simplification in Fluentd configuration needed for apiserver access logs after migrating to json format
+
+#### Story 3 - Administrator uses Elasticsearch to debug problem with workload
+
+TODO(serathius): Show examples of some queries to Elasticsearch
+
+### Performance
+
+We are expecting improving performance of logger library due ability to change logging library to require less memory and cpu, then current `klog`. More concerning is increase in volume of logs produced by components.
```

Comment:
```suggestion
We are expecting improving performance of logger library due ability to change logging library to require less memory and cpu, than current `klog`. More concerning is increase in volume of logs produced by components.
```

Reply:
> We are expecting improving performance of logger library

Do we have any data around the expected performance improvements? And I'd suggest to add some text detailing how we'll make sure and track possible performance regressions.